pennylane/optimize/rotoselect.py

```diff
@@ -0,0 +1,137 @@
+"""Rotoselect gradient free optimizer"""
+
+import pennylane as qml
+from pennylane import numpy as np
+
+
+class RotoselectOptimizer:
+    r"""Rotoselect gradient free optimizer.
+
+    The Rotoselect optimizer minimizes an objective function with respect to the parameters and
+    gates of a quantum circuit without the need of calculating the gradient of the function. The
+    algorithm works by updating the parameters :math:`\theta = \theta_1, \dots, \theta_D` and gate
+    choices :math:`R = R_1,\dots,R_D` one at a time according to a closed-form expression for the
+    optimal value of the :math:`d^{th}` parameter :math:`\theta^*_d` when the other parameters and
+    gate choices are fixed:
+
+    .. math:: \theta^*_d = \underset{\theta_d}{\text{argmin}}\left<H\right>_{\theta_d}
+              = -\frac{\pi}{2} - \text{arctan2}\left(2\left<H\right>_{\theta_d=0}
+              - \left<H\right>_{\theta_d=\pi/2} - \left<H\right>_{\theta_d=-\pi/2},
+              \left<H\right>_{\theta_d=\pi/2} - \left<H\right>_{\theta_d=-\pi/2}\right),
+
+    where :math:`\left<H\right>_{\theta_d}` is the expectation value of the objective function
+    optimized over the parameter :math:`\theta_d`. :math:`\text{arctan2}(x, y)` computes the
+    element-wise arc tangent of :math:`x/y` choosing the quadrant correctly, avoiding, in
+    particular, division-by-zero when :math:`y = 0`.
+
+    The algorithm is described in further detail in `Ostaszewski et al. (2019) <https://arxiv.org/abs/1905.09692>`_
+
+    Keyword Args:
+        possible_generators (list): List containing the possible ``pennylane.ops.qubit`` operators
+            that are allowed in the circuit. Default is the set of Pauli matrices :math:`\{X, Y, Z\}`.
+    """
+
+    def __init__(self, **kwargs):
+        self.possible_generators = kwargs.get("possible_generators", [qml.RX, qml.RY, qml.RZ])
+
+    def step(self, objective_fn, x, generators):
+        """Update x with one step of the optimizer.
+
+        Args:
+            objective_fn (function): the objective function for optimization
+            x (array): NumPy array containing the current values of the variables to be updated
+            generators (list): List containing ``pennylane.ops.qubit`` operators to be used in the
+                circuit and updated
+
+        Returns:
+            array: the new variable values :math:`x^{(t+1)}` as well as the new generators.
+        """
+        # make sure that x is an array
+        if np.ndim(x) == 0:
+            x = np.array([x])
+
+        for d, _ in enumerate(x):
+            x[d], generators[d] = self._find_optimal_params(objective_fn, x, generators, d)
+
+        return x, generators
+
+    def update_possible_generators(self, possible_generators):
+        """Update list of possible generators.
+
+        Args:
+            possible_generators (list): list of new possible generators to be used in optimization
+        """
+        self.possible_generators = possible_generators
```

Comment:
It was more thought of as an option for the user, if he/she wants to restrict the usage/optimization to certain gates (or use other gates for optimization rather than the basic Pauli gates). I think that a property might be better in that case; that makes more sense.